docs/proposals/resource-quota-scoping.md

```diff
@@ -0,0 +1,350 @@
+
+## Security Impact
+
+None.
+
+## End User Impact
+
+The `kubectl` commands that render quota should display its scopes.
+
+## Performance Impact
+
+This feature will make having more quota documents in a namespace
+more common in certain clusters.  This impacts the number of quota
+documents that need to be incremented during creation of an object
+in admission control.  It impacts the number of quota documents
+that need to be updated during controller loops.
+
+## Developer Impact
+
+None.
+
+## Alternatives
+
+This proposal initially enumerated a solution that leveraged a
+`FieldSelector` on a `ResourceQuota` document.  A `FieldSelector`
+grouped an `APIVersion` and `Kind` with a selector over its
+fields that supported set-based requirements.  It would have allowed
+a quota to track objects based on cluster defined attributes.
+
+For example, a quota could do the following:
+
+* match `Kind=Pod` where `spec.restartPolicy in (Always)`
+* match `Kind=Pod` where `spec.restartPolicy in (Never, OnFailure)`
+* match `Kind=Pod` where `status.qualityOfService in (BestEffort)`
+* match `Kind=Service` where `spec.type in (LoadBalancer)`
+ * see [#17484](https://github.com/kubernetes/kubernetes/issues/17484)
+
+Theoretically, it would enable support for fine-grained tracking
+on a variety of resource types.  While extremely flexible, there
+are cons to to this approach that make it premature to pursue
+at this time.
+
+* Generic field selectors are not yet settled art
+ * see [#1362](https://github.com/kubernetes/kubernetes/issues/1362)
+ * see [#19084](https://github.com/kubernetes/kubernetes/pull/19804)
+* Discovery API Limitations
+ * Not possible to discover the set of field selectors supported by kind.
+ * Not possible to discover if a field is readonly, readwrite, or immutable
+ post-creation.
+
+The quota system would want to validate that a field selector is valid,
+and it would only want to select on those fields that are readonly/immutable
+post creation to make resource tracking work during update operations.
+
+The current proposal could grow to support a `FieldSelector` on a
+`ResourceQuotaSpec` and support a simple migration path to convert
+`scopes` to the matching `FieldSelector` once the project has identified
+how it wants to handle `fieldSelector` requirements longer term.
+
+This proposal previously discussed a solution that leveraged a
+`LabelSelector` as a mechanism to partition quota.  This is potentially
+interesting to explore in the future to allow `namespace-admins` to
+quota workloads based on local knowledge.  For example, a quota
+could match all kinds that match the selector
+`tier=cache, environment in (dev, qa)` separately from quota that
+matched `tier=cache, environment in (prod)`.  This is interesting to
+explore in the future, but labels are insufficient selection targets
+for `cluster-administrators` to control footprint.  In those instances,
+you need fields that are cluster controlled and not user-defined.
```

Comment:
What is the downside of publishing every field as a label (in a separate namespace)? It would allow us to unify field selectors and label selectors (both in use and in implementation/code). If it's in a special namespace we could hide it in when displaying kubectl unless you ask to see that namespace. It would eliminate other kinds of redundancies too, for example having to publish QoS both as a field and as a label (latter so you can use it in MetadataPolicy, former for reasons @derekwaynecarr will remember but not me).